### ISSUE TITLE
Type narrowing bug in empty interfaces

### ISSUE THREAD

**Issue by chancancode** [Microsoft/TypeScript#18196]
[Reproduction](https://www.typescriptlang.org/play/#src=type%20Opaque%20%3D%20%7B%7D%20%7C%20null%20%7C%20undefined%20%7C%20void%3B%0D%0A%0D%0Afunction%20foo(value%3A%20Opaque)%3A%20void%20%7B%0D%0A%20%20if%20(value%20%7C%7C%20value%20%3D%3D%3D%200)%20%7B%0D%0A%20%20%20%20%2F%2F%20...%0D%0A%20%20%7D%20else%20%7B%0D%0A%20%20%20%20%2F%2F%20...%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0A%2F%2F%20You%20can%20see%20what%20is%20causing%20this%20bug%20here%3A%0D%0A%0D%0Ainterface%20Anything%20%7B%20%7D%0D%0A%0D%0Afunction%20bar(value%3A%20Anything)%20%7B%0D%0A%20%20if%20(value%20%7C%7C%20%2F*%20hover%20here%20-%3E%20*%2F%20value%20%2F*%20%3C-%20value%20is%20%22never%22%20*%2F%20%3D%3D%3D%200)%20%7B%20%0D%0A%0D%0A%20%20%7D%0D%0A%7D%0D%0A)

**Comment by DanielRosenwasser** [comment]
Is the proposal to always revert back to the declared type in a false branch when narrowing on truthiness?

**Comment by DanielRosenwasser** [comment]
Or perhaps to just not narrow in the false branch?

**Comment by wycats** [comment]
Maybe there's something we're misunderstanding about `{}`, but it doesn't seem like simply removing the falsy values from `{}` should result in `never`, right?

**Comment by chancancode** [comment]
> Is the proposal to always revert back to the declared type in a false branch when narrowing on truthiness?

I think the bug is that `{}` is incorrectly (IMO) considered "always truthy" which seems wrong to me as you can assign `0` or `""` to it just fine.

So today:

```ts
let x: string = ...;

if (!x) {
  // x is still `string`
}

let y: number = ...;

if (!y) {
  // y is still `number`
}
```

To the extent that ^^^ is the current behavior, I think `{}` should be treated the same, so, yes, I think `{}` should not be narrowed out in the false branches.

Ideally, since we have literally types now, I think all false branches should be something like `originalType & (0 | "" | undefined | null | false)`, so `x` in above would become `""` and `y` would become `0`, but that probably requires some machinery that is not in the system today.

**Comment by mhegazy** [comment]
Technically you are correct in the case of `{}`.  

Nevertheless, I think a better approach here is to use `object | string | number | boolean` instead of `{}` (I am assuming here that is what you originally intended)

**Comment by wycats** [comment]
![technically correct](http://i0.kym-cdn.com/photos/images/newsfeed/000/909/991/48c.jpg)

@mhegazy at minimum, your union is missing `symbol`.

```ts
const x: object | string | number | boolean = Symbol(); // does not type check
const x: {} = Symbol(); // type checks
```

It seems problematic for us to need to maintain a union of all of the possible primitive types ourselves.

From my perspective, there are a few different things we might be trying to express with these core types:

* "indexable" (you are allowed to use `[]` or `.x` at runtime)
  * anything other than `null` or `undefined`
  * hard to reliably express in TypeScript without enumerating a list of the primitive types
* "nullish"
  * `null | undefined`
  * `| void`?
* "falsy"
  * `null | undefined | false | 0 | ""`
* "truthy"
  * anything not falsy
  * afaict impossible to express in TypeScript

We typically try to wrangle these types when we start new projects in TypeScript, and we started to formalize our own usage in [ts-std](https://github.com/tildeio/ts-std). Our `core.ts` file has the types we use most often that aren't already in TypeScript: [ts-std/src/core.ts](https://github.com/tildeio/ts-std/blob/master/src/core.ts).

We are happy to use whatever unions work to express the most common concepts, and got the idea for `ts-std`'s `unknown` type from a very helpful discussion with @mhegazy early on in our TypeScript work.

We're happy to use `{}` to mean "not nullish", or happy to use a sanctioned union to mean that, but we really would like a canonical answer or implementation from the TypeScript team 😄

**Comment by mhegazy** [comment]
`symbol` is another primitive type. i would say `{}` is overly general in many cases. if i were starting over i would be more explicit, i.e. `type primitive = number | string | boolean | symbol | undefined | null | void | object` to mean `{} | null | undefined`.

I would like to fix this issue too, but will need to see what breaks. i am sure some one somewhere depends on this behavior inadvertently.

**Comment by DanielRosenwasser** [comment]
Side note: the list of falsy types needs to include: `NaN`.

**Comment by chancancode** [comment]
@danielrosenwasser that is correct. However, as far as I can tell, NaN is not a literal type in TypeScript. Is there a trick?

**Comment by mhegazy** [comment]
`NaN` is not a type. the value `NaN` is of type `number`. and `NaN` is funny from a typing perspective, so let's not go there right now..

**Comment by wycats** [comment]
To recap:

1. We agree that there is a bug with narrowing in `{}`. Specifically, just as `(! operation) & string = string` and `(! operation) & number = number`, `(! operation) & {}` should `= {}`.

2. @mhegazy proposed that we should just try to avoid using `{}`, and instead enumerate the types that `{}` is expressing. However, the list of those types is not permanently bounded. For example, `symbol` was recently introduced by JavaScript, which now must be included in such a list.

It seems that there are two options:

1. Fix the bug in `{}` and say that it means "the union of all not-null, not-undefined types"
2. Introduce a new built-in type that means "the union of all not-null, not-undefined types" that doesn't have the bug.

It is worth noting that JavaScript is adding a new primitive type ([BigInt](https://tc39.github.io/proposal-bigint/)) that has `typeof bigint === 'bigint'` and has `ToBoolean(0n) === false`, so there will be additional types that are subject to this problem over time.

**Comment by DanielRosenwasser** [comment]
Hey @chancancode and @wycats, now that `unknown` exists, I'd like to know whether you still end up running into this scenario.

**Comment by ljharb** [comment]
`unknown` still includes `null` and `undefined`, though; so it'd be nice if there was a simple way to refer to `object | string | boolean | number | symbol | bigint` etc.
